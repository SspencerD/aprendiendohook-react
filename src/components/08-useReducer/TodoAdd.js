import React from 'react'
import { useForm } from '../../Hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {

    const [{ description }, handleinputChange, reset] = useForm({
        description: ''

    });


    const handleSubmit =(e)=>{
        e.preventDefault();

        if (description.trim().length <= 1) {

            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAddTodo(newTodo);
        reset();

    }

    return (
        <>

            <h4>Agregar</h4>
            <hr />
            <form className="mb-3" onSubmit={handleSubmit}>
                <input
                    className="form-control form-control-sm"
                    type="text"
                    name="description"
                    placeholder="Aprender algo..."
                    autocomplete="off"
                    value={description}
                    onChange={handleinputChange}
                />
                <button
                    className="btn btn-sm btn-outline-success mt-2 btn-block"
                    type="submit"
                >
                    Agregar
                        </button>
            </form>

        </>
    )
}
