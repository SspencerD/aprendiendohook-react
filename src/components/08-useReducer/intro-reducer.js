
const initialState = [{
    id:1,
    todo: 'Estudiar mucho react',
    done: false
}];

const toDo = (state = initialState, action) => {

    if( action?.type=== 'agregar'){
        // como no se ocupa push , se puede ocupar el operador spread
        // ya que llama a la lista anterior y lo agrega
        return [...state , action.payload ];
    }
    //solo puede tener una sola condici√≥n , no se puede agregar un else
    // y otro return ya que estaria en contra del reducer

    return state;
}

let todos = toDo();

// No usar push si estan usando react (reducer)
// ya que modifica  o muta el obj

const newTodo ={
    id:2,
    todo: 'Estudiar mucho redux',
    done: false

}

const addAction ={

    // payload nombre de como se llama al nuevo objeto

    type: 'agregar',
    payload: newTodo
}

todos = toDo(todos, addAction );

console.log(todos);