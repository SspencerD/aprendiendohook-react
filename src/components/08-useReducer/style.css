body{
    padding: 50px;
}

p{
    cursor: pointer;
    margin: 0 px;
}

h1{
    align-items: center;
}

.list-group-item{
    display: flex;
    align-items: center;
    justify-content: space-between;
}
.badge{
    padding-right:  -10px;
    border-radius: 21px;
    cursor: pointer;
}

.complete{
    text-decoration: line-through;
}
