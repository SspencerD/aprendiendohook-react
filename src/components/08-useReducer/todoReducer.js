

export const todoReducer = (state = [], action) => {

    switch (action.type) {
        case 'add':

            return [...state, action.payload];

        case 'delete':

            return state.filter(todo => todo.id !== action.payload); // seria el grupo id

        case 'toggle': //metodo abreviado

        return state.map( todo =>
            ( todo.id === action.payload)
            ? {...todo, done: !todo.done}
            :todo );
        
            case 'toggle-old': //metodo muy largo de hacerlo

            return state.map(todo => {

                if (todo.id === action.payload) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                }else {
                    return todo;
                }

            
            })

        default:
return state;
    }


}