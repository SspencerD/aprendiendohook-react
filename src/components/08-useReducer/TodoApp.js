import React, { useEffect, useReducer } from 'react'
import './style.css';
import { todoReducer } from './todoReducer';
import { Todolist } from './Todolist';
import { TodoAdd } from './TodoAdd';



const init = () => {

    return JSON.parse(localStorage.getItem('todos')) || [];


    // Esto estamos haciendo al momento de renderizar el init
    // return [{
    //     id: new Date().getTime(),
    // desc: 'Aprender React',
    // done: false
    // }];

}
export const TodoApp = () => {


    const [todos, dispatch] = useReducer(todoReducer, [], init);


    useEffect(() => {

        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos]);

    const handleDelete = (todoId) => {

        const borrar = {
            type: 'delete',
            payload: todoId
        };

        dispatch(borrar);

    }

    const handleToggle = (todoId) => {

        dispatch({
            type: 'toggle',
            payload: todoId
        });
    }

    const handleAddTodo = (newTodo) => {

        dispatch({
            type: 'add',
            payload: newTodo

        });
    }



    return (
        <div>

            <h1 className="alert alert-dark">To Do APP <small className="badge bg-info"> {todos.length} </small> </h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <Todolist
                        todos={todos}
                        handleToggle={handleToggle}
                        handleDelete={handleDelete}
                    />
                </div>

                <div className="col-5">

                    <TodoAdd handleAddTodo={ handleAddTodo } />
                   
                </div>
            </div>
        </div>


    )
}
