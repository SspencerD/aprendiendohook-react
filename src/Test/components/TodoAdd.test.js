import { shallow } from 'enzyme';
import { TodoAdd } from '../../components/08-useReducer/TodoAdd'
import { demoTodo } from '../Data/datatodos';


describe('Pruebas en <TodoAdd/>', () => {

    const handleAddTodo = jest.fn();

    const wrapper = shallow(<TodoAdd

        handleAddTodo={handleAddTodo}
    />);

    test('debe de mostrarse correctamente', () => {

        expect(wrapper).toMatchSnapshot();
    });


    test('No debe de llamar a handleAddTodo', () => {

        // creamos una const llamando al la var wrapper en la cual debe buscar el form sobre la propiedad onsubmit
        const formsubmit = wrapper.find('form').prop('onSubmit');

        // llamamos la funcion formsubmit con el preventDefault
        formsubmit({ preventDefault() { } });


        // esperamos que llamada del handleAddTodo sea 0 ya que no queremos llamarlo
        expect(handleAddTodo).toHaveBeenCalledTimes(0);

    });


    test('Debe de llamar la funcíon handleAddTodo', () => {

        // creamos una variable X , para  incertarla luego en el arreglo
        const val = 'Aprender redux';

        // buscamos el input y simulamos el cambio que queremos realizarle ,
        // agregando que en el target debe haber un valor y una descripción
        wrapper.find('input').simulate('change', {
            target: {
                value: val,
                name: 'description'
            }
        });

        // creamos una constante del submit en la cual busca el formulario sobre la prop onSubmit
        const formsubmit = wrapper.find('form').prop('onSubmit');

        // llamamos la funcion formsubmit con el preventDefault
        formsubmit({ preventDefault() { } });

        // primer expresión seria llamar al componente una vez
        expect(handleAddTodo).toHaveBeenCalledTimes(1);
        // llamar al componente que esperando contenga algún objeto
        expect(handleAddTodo).toHaveBeenCalledWith( expect.any(Object) );

        // llamar al componente que contenga tal objeto según la var declarada ,
        // el done en falso y que la id contenga cualquier numero.
        expect(handleAddTodo).toHaveBeenCalledWith( {
            desc: val,
            done: false,
            id: expect.any(Number)
         });

         //llamar al componente esperando que una vez de la inseción de datos 
         //se reinicie el valor
         expect(wrapper.find('input').prop('value')).toBe('');
    });

});