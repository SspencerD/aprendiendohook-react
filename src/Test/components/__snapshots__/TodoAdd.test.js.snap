// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAdd/> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h4>
    Agregar
  </h4>
  <hr />
  <form
    className="mb-3"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control form-control-sm"
      name="description"
      onChange={[Function]}
      placeholder="Aprender algo..."
      type="text"
      value=""
    />
    <button
      className="btn btn-sm btn-outline-success mt-2 btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
